### PR TITLE
[4.x] disable external entity loader for libxml

### PR DESCRIPTION
**Pull request by SniperSister**
### Summary of Changes
Older versions of libxml and/or specific configurations are vulnerable to so called XML External Entity Processing vulnerabilities, short XXE:
https://owasp.org/www-community/vulnerabilities/XML_External_Entity_(XXE)_Processing

In order to harden J4 against such attacks, we should disable the loading of external entites by default.

### Testing Instructions
Test parts of the CMS related to XML processing i.e.:

* Parse RSS feeds using mod_feed
* Purge the extension cache in the extension manager
* Install an extension with an update server being enable

### Expected result
Features continue to work as expected.

### Documentation Changes Required
3rd party developers relying on external entity processing need to be aware of that change.

### CODE REVIEW

**Comment by richard67** [comment]
@SniperSister Please use tabs instead of spaces for indentation to fix PHPCS errrors mentioned by Drone.

**Comment by SniperSister** [comment]
@richard67 thanks, fixed!

File: includes/framework.php [hunk 1]
```diff
@@ -91,3 +91,13 @@
 }
 
 unset($config);
+
+// Suppress loading external XML entites to prevent XXE attacks
+if (function_exists('libxml_disable_entity_loader'))
+{
+	// disable XML external entity loading explicitly
```
Comment (richard67) [inline includes/framework.php]:
```suggestion
	// Disable XML external entity loading explicitly
```

File: includes/framework.php [hunk 2]
```diff
@@ -91,3 +91,13 @@
...
+	$intialValue = libxml_disable_entity_loader(true);
+
+	// and restore XML external entity loading on exit
```
Comment (richard67) [inline includes/framework.php]:
```suggestion
	// Restore XML external entity loading on exit
```

**Comment by richard67** [comment]
@SniperSister Drone still wasn't happy with PHPCS of comments, have corrected that for you.

**Comment by richard67** [comment]
@SniperSister I've just tested everything and a bit more than described, but I have a problem with langmetadata.xml. On the site (frontend) I get:

```
Warning: simplexml_load_file(): I/O warning : failed to load external entity "/joomla-cms-4.0-dev/language/en-GB/langmetadata.xml" in /joomla-cms-4.0-dev/libraries/src/Language/LanguageHelper.php on line 682
```

I don't really understand why simplexml finds an external entity in this file [https://github.com/joomla/joomla-cms/blob/4.0-dev/language/en-GB/langmetadata.xml](https://github.com/joomla/joomla-cms/blob/4.0-dev/language/en-GB/langmetadata.xml).

In the backend that doesn't happen.

**Comment by SniperSister** [comment]
@richard67 that's indeed super weird. Reproducible on multiple machines?

**Comment by richard67** [comment]
@SniperSister I have only one available right now, it's Linux.

**Comment by richard67** [comment]
PHP version 7.3.11-0ubuntu0.19.10.4

**Comment by wilsonge** [comment]
If we're doing this it needs to go in all 3 framework.php files for each of the applications

**Comment by carcam** [comment]
I have tested this item :white_check_mark: successfully on 8a3863757bac419a976565199bebdb8fcdfe1ff9<br /><br />I have tested this adding https://community.joomla.org/blogs/community.feed feed to the site and also installing a J4 extension and it's correctly processed in both cases.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28949">issues.joomla.org/tracker/joomla-cms/28949</a>.</sub>

**Comment by pabloarias** [comment]
I have tested this item :white_check_mark: successfully on 8a3863757bac419a976565199bebdb8fcdfe1ff9<br /><br />I've tested this path showing a RSS feed in frontend with News Feeds core component. After apply path it continues working.

I've installed a component too and purged the extensions cache in the extension manager, after search for updates, it works.

Thank you!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28949">issues.joomla.org/tracker/joomla-cms/28949</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28949">issues.joomla.org/tracker/joomla-cms/28949</a>.</sub>

**Comment by mbabker** [comment]
Not RTC.  George pointed out a required change, that has not been made.

**Comment by SharkyKZ** [comment]
I get the same warning as @richard67. PHP 7.2.5~7.4.6 on Windows.

**Comment by SharkyKZ** [comment]
`libxml_disable_entity_loader()` is deprecated in PHP 8.0.

**Comment by paritshivani** [comment]
I have tested this item :white_check_mark: successfully on 8a3863757bac419a976565199bebdb8fcdfe1ff9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28949">issues.joomla.org/tracker/joomla-cms/28949</a>.</sub>

**Comment by brianteeman** [comment]
@SniperSister @zero-24 as this is a security issue it would be good to get this resolved.

**Comment by zero-24** [comment]
I have just added the relevant changes to the administrator and api location as requested.

**Comment by wilsonge** [comment]
@richard67 can you still reproduce your error with the metadata?

**Comment by richard67** [comment]
> @richard67 can you still reproduce your error with the metadata?

@wilsonge Unfortunately yes:

2021-01-19_j4-pr-28949

I only made a new installation of current 4.0-dev, then switched error reporting to maximum (and server time zone to "Europe/Berlin", but I think that's not relevant here) in Global Configuration, then installed Patchtester 4, entered the GitHub token in Patchtester access information, then fetched the PRs in Patchtester, applied the patch of this PR, and then I got the above result.

From system info:
- PHP Version = 7.3.11-0ubuntu0.19.10.6
- Web Server = Apache/2.4.41 (Ubuntu)
- WebServer to PHP Interface = apache2handler

**Comment by laoneo** [comment]
I would close this as the function got deprecated in PHP 8.0, because it does this automatically as mentioned here https://php.watch/versions/8.0/libxml_disable_entity_loader-deprecation. If we really want to patch it, then I would also add a php check to execute the function only in PHP < 8.0.

**Comment by SniperSister** [comment]
Closed for the reasons mentioned by @laoneo

**Pull request closed without merging**